### PR TITLE
Refactor integration level so "no project opened" state is possible

### PR DESCRIPTION
**Pull request by vitvakatu**
### Pull Request Description

Code in `ide::integration` module refactored, to support `None` coming from `ide::Api::current_project()`.
`controller::searcher` now stores `project::Model` to avoid additional calls to `current_project()` (as they require unwraps now that looks suspicious). 

### Important Notes

Cloud environment require passing `project` parameter, as `controller::ide::plain` wasn't changed and does not support "no opened project" state. `controller::ide::desktop` however, allows this param not to be supplied. Project integration level will not be initialized in this case.

### Checklist

[ci no changelog needed]

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: app/gui/src/controller/ide/desktop.rs [hunk 1]
```diff
@@ -48,44 +48,49 @@ pub struct Handle {
...
-    /// Create a project controller handle with already loaded project model.
-    pub fn new_with_project(
+    /// Create IDE controller. If `maybe_project_name` is `Some`, a project with provided name will
+    /// be opened. Otherwise controller will be used for project manager operations by Welcome
+    /// Screen.
+    pub async fn new(
         project_manager: Rc<dyn project_manager::API>,
-        initial_project: model::Project,
-    ) -> Self {
+        maybe_project_name: Option<ProjectName>,
```
Comment (MichaelMauderer) [inline app/gui/src/controller/ide/desktop.rs]: I would just call it `project_name`. The "maybe" part is indicated by the type.

File: app/gui/src/controller/ide/desktop.rs [hunk 2]
```diff
@@ -48,44 +48,49 @@ pub struct Handle {
 }
 
 impl Handle {
-    /// Create a project controller handle with already loaded project model.
-    pub fn new_with_project(
+    /// Create IDE controller. If `maybe_project_name` is `Some`, a project with provided name will
+    /// be opened. Otherwise controller will be used for project manager operations by Welcome
+    /// Screen.
+    pub async fn new(
```
Comment (MichaelMauderer) [inline app/gui/src/controller/ide/desktop.rs:54]: Is there a reason this needs to be async now?

File: app/gui/src/ide/initializer.rs [hunk 3]
```diff
@@ -109,19 +106,17 @@ impl Initializer {
...
+                let controller =
+                    controller::ide::Desktop::new(project_manager, maybe_project_name).await?;
                 Ok(Rc::new(controller))
             }
             LanguageServer { json_endpoint, binary_endpoint, namespace } => {
                 let json_endpoint = json_endpoint.clone();
                 let binary_endpoint = binary_endpoint.clone();
                 let namespace = namespace.clone();
-                let project_name = self.config.project_name.clone();
+                // unwrap is safe here, as BackendService constructor made this check
```
Comment (MichaelMauderer) [inline app/gui/src/ide/initializer.rs]: Please write a full sentence in the comment. Also, it while this should never happen, it would still be better to use an `expect` with a meaningful error message in case a bug somewhere triggers it at some point.

File: app/gui/src/controller/searcher.rs [hunk 4]
```diff
@@ -477,6 +477,7 @@ pub struct Searcher {
     ide:              controller::Ide,
     this_arg:         Rc<Option<ThisNode>>,
     position_in_code: Immutable<TextLocation>,
+    project_name:     Immutable<project::QualifiedName>,
```
Comment (MichaelMauderer) [inline app/gui/src/controller/searcher.rs]: Is there a deeper reason why the searcher no stores the project name instead of deferring to the project? Does this account for renaming of the project?

**Review by MichaelMauderer — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher.rs]: Well, the only reason to store name is because otherwise it would require unwrapping, and it's hard to prove its safety on type level now. I didn't study searcher code and don't know whether it would affect renaming. @farmaazon as you proposed this change, do you have any idea?

I intend to revert this change and go with unwrapping instead.

File: app/gui/src/controller/ide/desktop.rs (hunk 2 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs:54]: It calls async function `Self::init_project_model`

File: app/gui/src/controller/ide/desktop.rs [hunk 5]
```diff
@@ -39,7 +39,7 @@ const UNNAMED_PROJECT_NAME: &str = "Unnamed";
 #[derivative(Debug)]
 pub struct Handle {
     logger:               Logger,
-    current_project:      Rc<CloneRefCell<model::Project>>,
+    current_project:      Rc<RefCell<Option<model::Project>>>,
```
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs]: Why this change?

File: app/gui/src/controller/ide/plain.rs [hunk 6]
```diff
@@ -83,8 +83,8 @@ impl Handle {
 }
 
 impl controller::ide::API for Handle {
-    fn current_project(&self) -> model::Project {
-        self.project.clone_ref()
+    fn current_project(&self) -> Option<model::Project> {
+        Some(self.project.clone_ref())
```
Comment (wdanilo) [inline app/gui/src/controller/ide/plain.rs:87]: I don't understand this implementation. It always returns `Some`, so what is the reason of it returning `Option` at all?

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (wdanilo) [reply app/gui/src/controller/searcher.rs]: Please talk with @farmaazon about proper design here. Probably after project is initialized, we should have a data structure like `InitializedProject` and searcher should have access to it. This structure should contain all project-related information for initialized project. I believe there is more of info than just name.

File: app/gui/src/ide/initializer.rs [hunk 7]
```diff
@@ -109,19 +106,17 @@ impl Initializer {
...
+                let project_name = self.config.project_name.clone().unwrap();
```
Comment (wdanilo) [inline app/gui/src/ide/initializer.rs]: Please refactor the code in such a way that it does not use unwrap. Making code "safe by comment" is very bad. Someone will refactor that, do not notice the comment and it will break. Do not use `unwrap` or other possible crashing functions until absolutely necessary.

File: app/gui/src/ide/integration.rs [hunk 8]
```diff
@@ -35,7 +35,13 @@ impl Model {
         // not race for the view.
         *self.project_integration.borrow_mut() = None;
 
-        let project_model = self.controller.current_project();
+        let project_model =
+            if let Some(model) = self.controller.current_project() { model } else { return };
```
Comment (wdanilo) [inline app/gui/src/ide/integration.rs]: do not use return. We are not using early returning functions inside of function bodies because it makes the code hard to follow. Use if-then-else instead

File: app/gui/tests/language_server.rs [hunk 9]
```diff
@@ -64,7 +64,7 @@ wasm_bindgen_test_configure!(run_in_browser);
 async fn ls_text_protocol_test() {
     let _guard = enso_gui::initializer::setup_global_executor();
     let ide = setup_ide().await;
-    let project = ide.current_project();
+    let project = ide.current_project().unwrap();
```
Comment (wdanilo) [inline app/gui/tests/language_server.rs]: don't use unwrap. I know these are tests, just check if its OK and do the test. If its None, fail the test with meaningful error.

**Comment by wdanilo** [comment]
Also, regarding:

> Cloud environment require passing project parameter, as controller::ide::plain wasn't changed and does not support "no opened project" state. controller::ide::desktop however, allows this param not to be supplied. Project integration level will not be initialized in this case.

This PR makes project state optional. If it makes it optional "in some places, but in other places not", then this makes the codebase not consistent. If we want to make the project optional, please fix in this PR all places where it is used, including cloud environment.

File: app/gui/src/controller/ide/desktop.rs (hunk 5 above)
Comment (farmaazon) [reply app/gui/src/controller/ide/desktop.rs]: > Why this change?

I can answer for @vitvakatu : The option does not implement CloneRef, only the immutable one. However this one is not immutable.

I could recommend the `CloneCell` - it is also in prelude.

File: app/gui/src/controller/ide/plain.rs (hunk 6 above)
Comment (farmaazon) [reply app/gui/src/controller/ide/plain.rs:87]: > I don't understand this implementation. It always returns `Some`, so what is the reason of it returning `Option` at all?

Because it is trait implementation. Other implementors can return `None`.

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs]: > Please talk with @farmaazon about proper design here. Probably after project is initialized, we should have a data structure like `InitializedProject` and searcher should have access to it. This structure should contain all project-related information for initialized project. I believe there is more of info than just name.

Ok, I suggested taking the project name, because this controller just does not need the whole project information. However, I forgot about renaming. 

We have already `InitializedProject` structure, and it's called `model::Project`. And I think this should be used here, the model might be requested for the project name (and it will work with renaming).

File: app/gui/src/ide/initializer.rs (hunk 7 above)
Comment (farmaazon) [reply app/gui/src/ide/initializer.rs]: If we require a project name when giving ls connection (and this makes sense), then the name should be put inside `BackendService::LanguageServer`

File: app/gui/src/ide/integration.rs (hunk 8 above)
Comment (farmaazon) [reply app/gui/src/ide/integration.rs]: @wdanilo I think this is not quite accurate, because we use a lot of `?` which are effectively return inside function body.

Still, at this place, +1 for making else block instead of return.

File: app/gui/tests/language_server.rs (hunk 9 above)
Comment (farmaazon) [reply app/gui/tests/language_server.rs]: > don't use unwrap. I know these are tests, just check if its OK and do the test. If its None, fail the test with meaningful error.

Perhaps `expect` method of Option fits here well.

File: app/gui/src/controller/ide/desktop.rs [hunk 10]
```diff
@@ -48,44 +48,49 @@ pub struct Handle {
...
 
 impl Handle {
-    /// Create a project controller handle with already loaded project model.
-    pub fn new_with_project(
+    /// Create IDE controller. If `maybe_project_name` is `Some`, a project with provided name will
+    /// be opened. Otherwise controller will be used for project manager operations by Welcome
+    /// Screen.
+    pub async fn new(
         project_manager: Rc<dyn project_manager::API>,
-        initial_project: model::Project,
```
Comment (farmaazon) [inline app/gui/src/controller/ide/desktop.rs:54]: I don't like removing the function taking initial project model; it could be handy in tests to provide already mocked model::Project without setting up mock of project_manager. Therefore, please:
1. keep the old `new_with_project` method
2. The `new` constructor should use new_with_project (as `new_with_opened_project did before).

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher.rs]: Yeah, I think a got a little too fancy here, we can store `model::Project`, and it is actually available in searcher's constructor.

**Comment by vitvakatu** [comment]
@wdanilo 
> This PR makes project state optional. If it makes it optional "in some places, but in other places not", then this makes the codebase not consistent. If we want to make the project optional, please fix in this PR all places where it is used, including cloud environment.

We agreed that the cloud won't support Welcome Screen in its first iteration, as it uses own project management UI. The codebase is still consistent, as project state becomes optional everywhere. Before we couldn't open IDE without a project - now we can. If cloud won't pass optional `project` param - welcome screen will open (and will probably not work). It is cloud's responsibility to always pass `project`, and it follows its current implementation. So I don't think it's possible to "fix in this PR all places where it is used".

File: app/gui/src/ide/integration.rs (hunk 8 above)
Comment (wdanilo) [reply app/gui/src/ide/integration.rs]: `?` and `return` are similar indeed. I would still ban usage of `return` and allow usage of `?`.

**Comment by wdanilo** [comment]
> @wdanilo
> 
> > This PR makes project state optional. If it makes it optional "in some places, but in other places not", then this makes the codebase not consistent. If we want to make the project optional, please fix in this PR all places where it is used, including cloud environment.
> 
> We agreed that the cloud won't support Welcome Screen in its first iteration, as it uses own project management UI. The codebase is still consistent, as project state becomes optional everywhere. Before we couldn't open IDE without a project - now we can. If cloud won't pass optional `project` param - welcome screen will open (and will probably not work). It is cloud's responsibility to always pass `project`, and it follows its current implementation. So I don't think it's possible to "fix in this PR all places where it is used".

fair enough, thanks for the explanation!

File: app/gui/src/controller/ide/desktop.rs (hunk 1 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: Fixed

File: app/gui/src/ide/initializer.rs (hunk 3 above)
Comment (vitvakatu) [reply app/gui/src/ide/initializer.rs]: Noted, comment removed as part of further refactoring: `project_name` now is a required part of `LanguageServer`, so no unwrap is needed.

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (vitvakatu) [reply app/gui/src/controller/searcher.rs]: Refactored to storing `model::Project`

File: app/gui/src/controller/ide/desktop.rs (hunk 5 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs]: Changed it to `CloneCell`, it fits nicer indeed.

File: app/gui/src/ide/initializer.rs (hunk 7 above)
Comment (vitvakatu) [reply app/gui/src/ide/initializer.rs]: Done as proposed, now project name is a part of `LanguageServer`, unwrap removed.

File: app/gui/src/ide/integration.rs (hunk 8 above)
Comment (vitvakatu) [reply app/gui/src/ide/integration.rs]: Fixed

File: app/gui/tests/language_server.rs (hunk 9 above)
Comment (vitvakatu) [reply app/gui/tests/language_server.rs]: Replaced with `expect`

File: app/gui/src/controller/ide/desktop.rs (hunk 10 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs:54]: I added `new_with_project_model` constructor (though it accepts `Option<model::Project>` or code reuse with `new` won't be possible).

**Review by farmaazon — APPROVED** [review]

**Review by MichaelMauderer — APPROVED** [review]

File: app/gui/tests/language_server.rs [hunk 11]
```diff
@@ -64,7 +64,7 @@ wasm_bindgen_test_configure!(run_in_browser);
 async fn ls_text_protocol_test() {
     let _guard = enso_gui::initializer::setup_global_executor();
     let ide = setup_ide().await;
-    let project = ide.current_project();
+    let project = ide.current_project().expect("IDE is configured without an open project");
```
Comment (mwu-tow) [inline app/gui/tests/language_server.rs]: We try to keep errors messages as nicely formatted sentences, ending with a dot.

**Review by mwu-tow — APPROVED** [review]

File: app/gui/src/controller/ide/desktop.rs (hunk 2 above)
Comment (wdanilo) [reply app/gui/src/controller/ide/desktop.rs:54]: @vitvakatu please not mark comments as resolved if you were not the original author of a comment. It is the role of the person who was asking questions to mark a comment as resolved. This should be left open until @MichaelMauderer closes it.

File: app/gui/src/controller/ide/desktop.rs [hunk 12]
```diff
@@ -48,13 +48,28 @@ pub struct Handle {
...
+        project_name: Option<ProjectName>,
+    ) -> FallibleResult<Self> {
+        let project = match project_name {
+            Some(name) => Some(Self::init_project_model(project_manager.clone_ref(), name).await?),
+            None => None,
```
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs:60]: This is a mapping, it should use map

File: app/gui/src/controller/ide/desktop.rs [hunk 13]
```diff
@@ -122,8 +136,9 @@ impl ManagingProjectAPI for Handle {
                 self.project_manager.create_project(&name, &version, &action).await?;
             let new_project_id = create_result.project_id;
             let project_mgr = self.project_manager.clone_ref();
-            let new_project = Project::new_opened(&self.logger, project_mgr, new_project_id);
-            self.current_project.set(new_project.await?);
+            let new_project =
+                Project::new_opened(&self.logger, project_mgr, new_project_id).await?;
```
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs]: Are you able to refactor this line in such a way that it is not broken into 2 lines? You can create some additional variables if needed. Our code style tells about a strong preference towards single line expressions over multi-line ones.

File: app/gui/src/controller/ide/desktop.rs [hunk 14]
```diff
@@ -142,8 +157,9 @@ impl ManagingProjectAPI for Handle {
         async move {
             let logger = &self.logger;
             let project_mgr = self.project_manager.clone_ref();
-            let new_project = model::project::Synchronized::new_opened(logger, project_mgr, id);
-            self.current_project.set(new_project.await?);
+            let new_project =
+                model::project::Synchronized::new_opened(logger, project_mgr, id).await?;
```
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs]: Are you able to refactor this line in such a way that it is not broken into 2 lines? You can create some additional variables if needed. Our code style tells about a strong preference towards single line expressions over multi-line ones.

File: app/gui/src/ide/initializer.rs [hunk 15]
```diff
@@ -109,19 +106,16 @@ impl Initializer {
...
                 let project_manager = self.setup_project_manager(endpoint).await?;
-                let project_name = self.config.project_name.clone();
-                let controller = controller::ide::Desktop::new_with_opened_project(
-                    project_manager,
-                    project_name,
-                )
-                .await?;
+                let maybe_project_name = self.config.project_name.clone();
+                let controller =
+                    controller::ide::Desktop::new(project_manager, maybe_project_name).await?;
```
Comment (wdanilo) [inline app/gui/src/ide/initializer.rs]: Are you able to refactor this line in such a way that it is not broken into 2 lines? You can create some additional variables if needed. Our code style tells about a strong preference towards single line expressions over multi-line ones.

**Review by wdanilo — APPROVED** [review]

File: app/gui/tests/language_server.rs (hunk 11 above)
Comment (vitvakatu) [reply app/gui/tests/language_server.rs]: 🙈 Always forget about that

File: app/gui/src/controller/ide/desktop.rs (hunk 12 above)
Comment (vitvakatu) [reply app/gui/src/controller/ide/desktop.rs:60]: It is actually impossible here due to async call to `init_project_model`, as you can't use `await` in `Option::map`'s closure. There might be one other solution through `futures::stream::iter` and `StreamExt::then`, but I think it's a little overkill.

**Comment by vitvakatu** [comment]
@mwu-tow @wdanilo codestyle fixes applied

File: app/gui/src/config.rs [hunk 16]
```diff
@@ -67,7 +72,16 @@ impl BackendService {
                         .namespace
                         .clone()
                         .unwrap_or_else(|| constants::DEFAULT_PROJECT_NAMESPACE.to_owned());
-                    Ok(Self::LanguageServer { json_endpoint, binary_endpoint, namespace })
+                    let project_name = args
+                        .project
+                        .clone()
+                        .ok_or_else(|| MissingOption(args.names().project()))?;
```
Comment (wdanilo) [inline app/gui/src/config.rs]: I have comment about that in other PRs. Please check such places in the code. We prefer multiple variables over multi-line expressions. I've been writing longer comments about that in other places. See the style guide for more info.

**Review by wdanilo — COMMENTED** [review]
I'm not accepting it yet, as there are things that are not fixed although they were requested to be fixed in the past. It is important to apply the requested reviews to the code.

**Comment by MichaelMauderer** [comment]
> I'm not accepting it yet, as there are things that are not fixed although they were requested to be fixed in the past. It is important to apply the requested reviews to the code.

@wdanilo I was just about to merge this. You had already accepted this PR. Could you clarify whether this is accepted or not and update your review status accordingly?

File: app/gui/src/config.rs (hunk 16 above)
Comment (vitvakatu) [reply app/gui/src/config.rs]: Apparently it's too easy to skip such places when briefly looking through the code, I fixed it.
